utils/stoppable_workers.go

```diff
@@ -31,23 +33,34 @@ type stoppableWorkersImpl struct {
 func NewStoppableWorkers(funcs ...func(context.Context)) StoppableWorkers {
 	cancelCtx, cancelFunc := context.WithCancel(context.Background())
 	workers := &stoppableWorkersImpl{cancelCtx: cancelCtx, cancelFunc: cancelFunc}
-	workers.activeBackgroundWorkers.Add(len(funcs))
+	workers.AddWorkers(funcs...)
+	return workers
+}
+
+// AddWorkers starts up additional goroutines for each function passed in.
+func (sw *stoppableWorkersImpl) AddWorkers(funcs ...func(context.Context)) {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
```

Comment:
Zack had a great suggestion: check if the context here, and if it's already canceled, exit early without spinning up any new goroutines.